posts/capture-data-history-with-scd2-using-databricks-delta-live-tables.md

```diff
@@ -0,0 +1,84 @@
+
+### Daily estimates for APPL:
+
+![](https://raw.githubusercontent.com/Falydoor/blog-usa/blog-scd2-with-dlt/images/2023/01/dlt-scd2-1.png)
+
+- `TICKER` for the stock ticker
+- `ESTIMATE_DATE` for the date of the estimation
+- `PRICE` for the price estimation
+- `NUMBER_OF_ESTIMATORS` for how many estimators were used to calculate the price
+- `FILE_DATE` for the date associated with the raw file
+- `ROW_HASH` for a hash calculated using all columns except `FILE_DATE` (to simplify comparing rows)
+
+### Daily estimates for APPL with SCD2:
+
+![](https://raw.githubusercontent.com/Falydoor/blog-usa/blog-scd2-with-dlt/images/2023/01/dlt-scd2-2.png)
+
+You can see that 3 columns were added: `START_DATE`, `END_DATE` and `IS_CURRENT`. Those are the columns that will give you the history so you can query row-level change information.
+
+For more details about all types of SCD, I recommend reading this [great blog](https://www.sqlshack.com/implementing-slowly-changing-dimensions-scds-in-data-warehouses/).
+
+# Retain history on your tables with DLT
+
+To capture your data changes and retain history with DLT, you will just need to use the function `dlt.apply_changes()` and provide the correct parameters. Those parameters are very important and are gonna be unique for each table. For more information about those arguments, please read [the documentation of apply_changes()](https://docs.databricks.com/workflows/delta-live-tables/delta-live-tables-cdc.html#apply-changes-function). Make also sure that your DLT pipeline has the configuration `pipelines.enableTrackHistory` set to `true` and that the edition is either `Pro` or `Advanced`.
+
+Based on the dataset from above, here how it would look:
+
+```python
+import dlt
+from pyspark.sql.functions import col, expr
+
+@dlt.view
+def stock_estimate():
+  return spark.readStream.format("delta").table("stock_estimate_raw")
+
+dlt.create_streaming_live_table("stock_estimate_scd2")
+
+dlt.apply_changes(
+  target = "stock_estimate_scd2",          # The target table
+  source = "stock_estimate",               # The temp view
+  keys = ["ticker", "estimate_date"],      # Columns to uniquely identify a row
+  sequence_by = "file_date",               # Logical order of the rows
+  stored_as_scd_type = "2",                # To use SCD2
+  track_history_column_list = ["row_hash"] # To track changes of the whole row
+)
+```
+
+With very low changes to your DLT pipeline, the `stock_estimate_scd2` table will now track history by adding 2 new columns `__START_AT`/`__END_AT` and they can be queried to know the validity period of the row.
+
+# Usage in the finance world
+
+The first kind of data that your data lake should have in order to have your Portfolio Managers happy are financials and filings data. Bloomberg data is usually the go to for that and your ELT process just has to process the data incrementally (since there is no estimate).
```

Comment:
```suggestion
To best support your portfolio managers, it's recommended that your data lake include financial and filings data. Bloomberg data is usually the best for that and your ELT process just has to process the data incrementally (since there is no estimate).
```